apps/tunes/utils.py

```diff
@@ -35,9 +36,20 @@ def generate_browse_playlist(
     :param artist: (str) Optional artist of songs to return
     :param jitter: (float) Optional "shuffle" for the boundary box to give users songs from outside their norm
     :param songs: (QuerySet) Optional queryset of songs to filter
+    :param user_auth: (SpotifyUserAuth) Optional auth instance for checking Spotify for user saved tracks
 
     :return playlist: (QuerySet) `QuerySet` of `Song` instances for the given parameters
     """
+
+    # Check if user has any songs from Spotify saved to their record
+    # If they do, return that first; we want the songs the user has
+    # saved to Spotify to appear first
+    if user_auth and user_auth.saved_songs:
+        user_playlist = songs.filter(code__in=user_auth.saved_songs)
```

Comment:
This is an incredibly unperformant query for a large volume of saved songs. How can we optimize this query?

Reply:
Refactored to doing filtering in Python, seems to be more performant